evm/db/account.py

```diff
@@ -115,37 +118,33 @@ def account_is_empty(self, address):
     # Record and discard API
     #
     @abstractmethod
-    def record(self) -> UUID:
+    def record(self) -> Tuple[UUID, UUID]:
         raise NotImplementedError("Must be implemented by subclass")
 
     @abstractmethod
-    def discard(self, checkpoint: UUID) -> None:
+    def discard(self, changeset: Tuple[UUID, UUID]) -> None:
         raise NotImplementedError("Must be implemented by subclass")
 
     @abstractmethod
-    def commit(self, checkpoint: UUID) -> None:
-        raise NotImplementedError("Must be implemented by subclass")
-
-    @abstractmethod
-    def clear(self) -> None:
+    def commit(self, changeset: Tuple[UUID, UUID]) -> None:
         raise NotImplementedError("Must be implemented by subclass")
 
 
 class AccountDB(BaseAccountDB):
 
     def __init__(self, db, state_root=BLANK_ROOT_HASH):
-        # Keep a reference to the original db instance to use it as part of _get_account()'s cache
-        # key.
-        self._unwrapped_db = db
-        self.db = JournalDB(db)
-        self._trie = HashTrie(HexaryTrie(self.db, state_root))
+        self._journaldb = JournalDB(db)
+        self._trie = HashTrie(HexaryTrie(db, state_root))
+        self._trie_cache = CacheDB(self._trie)
+        self._journaltrie = JournalDB(self._trie_cache)
```

Comment:
I could use comments/docstring/something here that explains what each of these is for.

Reply:
This is where I'm at for the docstring so far:

```
        r"""
        Internal implementation details (subject to rapid change): 
        Database entries go through several pipes, like so... 

        .. code:: 

                                                     -> hash-trie -> storage lookups
                                                    /
            db -----------------------> _journaldb ----------------> code lookups
             \
              > _trie -> _trie_cache -> _journaltrie --------------> account lookups

        Journaling sequesters writes here ^, until persist is called.

        _journaldb is a journaling of the keys and values used to store
        code and account storage.

        _trie is a hash-trie, used to generate the state root

        _trie_cache is a cache tied to the state root of the trie. It
        is important that this cache is checked *after* looking for
        the key in _journaltrie, because the cache is only invalidated
        after a state root change.

        _journaltrie is a journaling of the accounts (an address->rlp mapping,
        rather than the nodes stored by the trie). This enables
        a squashing of all account changes before pushing them into the trie.

        .. NOTE:: There is an opportunity to do something similar for storage

        AccountDB synchronizes the snapshot/revert/persist of both of the
        journals. 

        """
```